Wrap DateTimeImmutable instead of extending it

othercorey opened the pull request:
closes https://github.com/cakephp/chronos/issues/79

othercorey committed:
Change Chronos and ChronosDate to wrap DateTimeImmutable instead of extend it

othercorey commented on src/Chronos.php:
```diff
@@ -182,6 +196,112 @@ public static function hasTestNow(): bool
+    {
+        $native = $this->native->modify($modifier);
+        if ($native === false) {
+            throw new InvalidArgumentException('Unable to modify date using: ' . $modifier);
```
This throws an exception on error now.

othercorey commented on src/Traits/FrozenTimeTrait.php:
```diff
@@ -132,26 +145,35 @@ public function setTimezone(DateTimeZone|string $value): static
+        $new = clone $this;
+        $new->native = $new->native->modify($modifier);
+        if ($new->native === false) {
+            throw new InvalidArgumentException('Unable to modify date using: ' . $modifier);
```
This throws an exception on error now

othercorey commented on tests/TestCase/Date/TimezoneTest.php:
```diff
@@ -27,10 +27,8 @@ class TimezoneTest extends TestCase
     {
-        $tz = new DateTimeZone('Pacific/Honolulu');
         $date = new Date('2015-01-01');
-        $new = $date->setTimezone($tz);
```
`Date::setTimezone()` was incorrectly returning `$this` instead of a new instance.

ADmad commented on src/Chronos.php:
```diff
@@ -81,46 +83,59 @@ class Chronos extends DateTimeImmutable implements ChronosInterface
         }
 
-        if ($time instanceof DateTimeInterface) {
+        if (is_object($time)) {
```
The `Date` class will also have a timezone?

othercorey replied:
It currently is set to the default time zone.

ADmad commented on src/Chronos.php:
```diff
@@ -182,6 +196,112 @@ public static function hasTestNow(): bool
+     * @param int $microseconds Microseconds of the time
+     * @return static
+     */
+    public function setTime(int $hours, int $minutes, int $seconds = 0, int $microseconds = 0): static
```
Can we have all arguments default to 0? That would allow one to do for e.g. `->setTime(minutes: 1)`.

othercorey replied:
We can. This matched DateTimeImmutable::setTime(). I'm not sure why it was defaulting to null and not 0.

ADmad commented on src/Chronos.php:
```diff
@@ -182,6 +196,112 @@ public static function hasTestNow(): bool
+     * @param string $modifier Date modifier
+     * @return static
+     * @throws \InvalidArgumentException
+     */
```
A `@see` or `@link` tag to the PHP manual for the modifier string format would be nice.

ADmad commented on src/Chronos.php:
```diff
@@ -182,6 +196,112 @@ public static function hasTestNow(): bool
+     */
+    public function getTimezone(): DateTimeZone|false
+    {
+        return $this->native->getTimezone();
```
Are you aware in what case this would return `false`? Manual just says "...false in case of failure".

othercorey replied:
I could not figure it out. If you look at the PHP source, it is false if some internal timelib flag "is local time" is set. I have no idea how that happens.

ADmad replied:
I would change the return type for our method to just `DateTimeZone` and throw an exception if the native class returns false.

markstory replied:
Being more strict sounds good to me too.

othercorey commented:
Dropped ChronosInterface and moved all constants to Chronos. Moved all static configs to Chronos since they should be the same for all instances.

Did not move **all** trait functions into Chronos and ChronosDate only the ones that seem to require type changes. Hopefully that makes it easier to review.

ADmad commented on composer.json:
```diff
@@ -50,10 +50,12 @@
+            "@psalm"
         ],
-        "stan-setup": "cp composer.json composer.backup && composer require --dev phpstan/phpstan:~1.8.0 && mv composer.backup composer.json",
+        "stan-setup": "cp composer.json composer.backup && composer require --dev phpstan/phpstan:~1.8.0 psalm/phar:~5.0-beta && mv composer.backup composer.json",
```
We should just drop this now that we use the common workflows for CI and the versions used here will keep getting out of sync.

othercorey replied:
I just need something to install locally.

ADmad commented on src/Traits/FrozenTimeTrait.php:
```diff
@@ -33,13 +35,15 @@ trait FrozenTimeTrait
-    protected function stripTime(DateTime|DateTimeImmutable|string|int|null $time, ?DateTimeZone $timezone): string
-    {
+    protected function stripTime(
+        Chronos|ChronosDate|DateTime|DateTimeImmutable|string|int|null $time,
```
```suggestion
        Chronos|ChronosDate|DateTimeInterface|string|int|null $time,
```

ADmad commented on src/Traits/FrozenTimeTrait.php:
```diff
@@ -33,13 +35,15 @@ trait FrozenTimeTrait
      * Used to ensure constructed objects always lack time.
      *
-     * @param \DateTime|\DateTimeImmutable|string|int|null $time The input time. Integer values will be assumed
+     * @param \Cake\Chronos\Chronos|\Cake\Chronos\ChronosDate|\DateTime|\DateTimeImmutable|string|int|null $time The input time. Integer values will be assumed
```
```suggestion
     * @param \Cake\Chronos\Chronos|\Cake\Chronos\ChronosDate|\DateTimeInterface|string|int|null $time The input time. Integer values will be assumed
```

othercorey commented on tests/TestCase/Date/DiffTest.php:
```diff
@@ -0,0 +1,375 @@
+            return $date->dayOfWeek === 1;
+        }, $dt->endOfMonth()));
+    }
+    */
```
These disabled tests just happens to pass with Chronos because we use the current time instead of setting to 00:00:00. DatePeriod does not include the end date and the option to enable that was added in PHP 8.2.

othercorey replied:
Think I can work around this in 8.1 by adding time to the end date.

markstory replied:
Having date operations behave differently based on the time of day they are run is unexpected. Date operations should behave like calendar dates do.

I'm happy to do a follow up pull request to improve how this works.

markstory commented on src/Chronos.php:
```diff
@@ -183,6 +297,46 @@ public static function hasTestNow(): bool
+    /**
+     * Get the difference formatter instance or overwrite the current one.
+     *
+     * @param \Cake\Chronos\DifferenceFormatterInterface|null $formatter The formatter instance when setting.
+     * @return \Cake\Chronos\DifferenceFormatterInterface The formatter instance.
```
We could have a cake adapter for this interface that handles the translation aspects. That could be a way to resolve the 'how to align the cake and chronos feature set'. We align the formatting in the new version of Chronos, and then cake would provide the i18n integration.

markstory commented on src/Chronos.php:
```diff
@@ -247,6 +401,552 @@ public static function createInterval(
+     */
+    public function diff(Chronos|DateTimeInterface $target, bool $absolute = false): DateInterval
+    {
+        $target = $target instanceof DateTimeInterface ? $target : $target->native;
```
Love this tech of using PHP's scoping rules to access other objects :clap:

markstory commented on src/Chronos.php:
```diff
@@ -247,6 +401,552 @@ public static function createInterval(
+     */
+    public function greaterThan(Chronos $other): bool
+    {
+        return $this > $other;
```
```suggestion
        return $this->native > $other->native;
```

markstory commented on src/ChronosDate.php:
```diff
@@ -115,8 +142,416 @@ public function __construct(
+     *
+     * @return \DateTimeZone|false
+     */
+    public function getTimezone(): DateTimeZone|false
+    {
+        return $this->native->getTimezone();
+    }
```
I don't think dates should have Timezones. This was a mistake in the previous design that we could correct now. I'm happy to do that as a follow up pull request if you'd prefer.

othercorey replied:
I was just re-creating the api from the interface initially, but we should drop this and see what needs to be fixe in cake core.

othercorey commented:
Dropped time zone from ChronosDate and any functions affected by time zone. Dropped creating `ChronosDate()` from now.

Dropped:

* set/getTimezone
* getOffset
* set/getTimestamp
* now
* instance (same as parse)
* today/tomorrow
* isFuture/isPast/isNextWeek
* Related magic getters

Users can get the current date through Chronos: `new ChronosDate(Chronos::now('America/Toronto'))`

Moved many functions into Chronos and Chronos date to allow changes in implementation.

othercorey commented:
I also dropped `createFromDate()` since `create()` is `createFromDate()` for `ChronosDate`. I don't know if we want to keep that duplicate.

markstory commented on docs/ja/index.rst:
```diff
@@ -286,7 +286,7 @@ Chronos は、出力した日時オブジェクトを表示するための多く
     Chronos::setTestNow(Chronos::now());
     MutableDateTime::setTestNow(MutableDateTime::now());
-    ChronosDate::setTestNow(ChronosDate::now());
+    ChronosDate::setTestNow(ChronosDate::parse(Chronos::now()));
```
I don't love this API but its something we can simplify after this merges.

othercorey replied:
This feels like some sort of search and replace error. It should just be now().

markstory reviewed (approved):
Looks good to me. Lets merge it.
